PR: https://github.com/odoo/odoo/pull/253649

From: bb56600bfc81b9a6573a867df93a706de8154d0a
From: Moamen(msho)
Date: 2026-04-17 20:38:55

Breaking data model changes scores: del:1 + add:0, change matches:
-    manual_consumption = fields.Boolean(

Total Changes: 103

[IMP] mrp: remove manual consumption logic

This commit removes the technical field `manual_consumption` logic from
`mrp/stock_move`.

Before this commit:
If a move is marked as `manual_consumption=True`, its consumed quantity
is not automatically updated on the MO. `manual_consumption` logic
basically depended on the move being linked to an operation on the
bom_line_id.

After this commit:
No more `manual_consumption` field. A move's consumed quantity is not
automatically updated if it's picked.
How the move is being picked is the same which is by marking a
move as consumed in the shopfloor app, marking a work order
that has component moves as done or simply editing the consumed
quantity on the MO form or barcode.

Changes in behavior:
* If the move is marked as picked => it's no longer automatically
updated
* Using barcode for mrp: When updating the qty_producing on the finished
product, exiting the barcode and entering again => the quantity on the
move lines reflect the new changed qty_producing and no more unpicked
splitted move lines are created. However, changing the quantity on the
move Lines (instead of the qty_producing) causes the move lines to split
as expected [this is not necessarily a limitation but it can be
imporoved if needed].

Enterprise PR: https://github.com/odoo/enterprise/pull/110503
Upgrade PR: https://github.com/odoo/upgrade/pull/9794

closes odoo/odoo#253649

Task: 5882264
Related: odoo/enterprise#110503
Related: odoo/upgrade#9794
Signed-off-by: Tiffany Chang (tic) <tic@odoo.com>

================================= pseudo patch: =================================

--- a/addons/mrp/models/mrp_production.py
+++ b/addons/mrp/models/mrp_production.py
@@ -1433,7 +1433,6 @@ class MrpProduction(models.Model):
             'warehouse_id': source_location.warehouse_id.id,
             'reference_ids': self.reference_ids.ids,
             'propagate_cancel': self.propagate_cancel,
-            'manual_consumption': self.env['stock.move']._determine_is_manual_consumption(bom_line),
         }
         return data
 
@@ -1447,7 +1446,7 @@ class MrpProduction(models.Model):
     def _mark_byproducts_as_produced(self):
         self.move_byproduct_ids.picked = True
 
-    def _set_qty_producing(self, pick_manual_consumption_moves=True):
+    def _set_qty_producing(self, mark_moves_picked=True):
         if self.product_id.tracking == 'serial':
             qty_producing_uom = self.uom_id._compute_quantity(self.qty_producing, self.product_id.uom_id, rounding_method='HALF-UP')
             qty_production_uom = self.uom_id._compute_quantity(self.product_qty, self.product_id.uom_id, rounding_method='HALF-UP')
@@ -1463,17 +1462,14 @@ class MrpProduction(models.Model):
             | self.move_finished_ids.filtered(lambda m: m.product_id != self.product_id or m.product_id.tracking == 'serial')
         ):
             is_byproduct = move in self.move_byproduct_ids
-            # Never update already produced by-product moves.
-            if move.picked and (is_byproduct or move.manual_consumption):
-                continue
-
+            # Never update already picked moves.
             # sudo needed for portal users
-            if move.sudo()._should_bypass_set_qty_producing():
+            if move.picked or move.sudo()._should_bypass_set_qty_producing():
                 continue
 
             new_qty = move.uom_id.round((self.qty_producing - self.qty_produced) * move.unit_factor)
             move._set_quantity_done(new_qty)
-            if (not move.manual_consumption or pick_manual_consumption_moves) \
+            if mark_moves_picked \
                     and move.quantity \
                     and not is_byproduct \
                     and (move.raw_material_production_id or move.product_id.tracking != 'serial'):
@@ -1869,7 +1865,8 @@ class MrpProduction(models.Model):
                 elif line := move.bom_line_id:
                     expected_qty = line.uom_id._compute_quantity(
                         line.product_qty * bom_factors[line.bom_id.id], move.uom_id)
-                picked_qty = move._get_picked_quantity()
+                picked_qty = sum(ml.uom_id._compute_quantity(ml.quantity, move.uom_id, round=False)
+                    for ml in move.move_line_ids)
                 if move.uom_id.compare(picked_qty, expected_qty):
                     issues.append((order, move.product_id, move.uom_id, move, picked_qty, expected_qty))
             # Once we've matched every orphaned BoM line we could match with orphaned moves, we consider
@@ -2304,6 +2301,9 @@ class MrpProduction(models.Model):
         backorders = productions_to_backorder and productions_to_backorder._split_productions()
         backorders = backorders - productions_to_backorder
 
+        for production in self.filtered(lambda p: not p.uom_id.is_zero(production.qty_producing)):
+            production.move_raw_ids.filtered(lambda m: not m.picked).picked = True
+
         productions_not_to_backorder._post_inventory(cancel_backorder=True)
         productions_to_backorder._post_inventory(cancel_backorder=True)
 
@@ -2413,12 +2413,7 @@ class MrpProduction(models.Model):
         self._button_mark_done_sanity_checks()
         production_auto_ids = set()
         for production in self:
-            if not production.uom_id.is_zero(production.qty_producing):
-                production.move_raw_ids.filtered(
-                    lambda move: move.manual_consumption and not move.picked
-                ).picked = True
-                continue
-            if production._auto_production_checks():
+            if production.uom_id.is_zero(production.qty_producing) and production._auto_production_checks():
                 production_auto_ids.add(production.id)
 
         productions_auto = self.env['mrp.production'].browse(production_auto_ids)
@@ -2790,7 +2785,6 @@ class MrpProduction(models.Model):
                 if move_raw.operation_id != bom_line.operation_id:
                     move_raw.operation_id = bom_line.operation_id
                     move_raw.workorder_id = self.workorder_ids.filtered(lambda wo: wo.operation_id == move_raw.operation_id)
-                move_raw.manual_consumption = move_raw._determine_is_manual_consumption(bom_line)
             elif not bom_line:
                 moves_to_unlink |= move_raw
         # Creates a raw moves for each remaining BoM's lines.
@@ -3019,9 +3013,9 @@ class MrpProduction(models.Model):
         for move in self.move_raw_ids:
             if move.state in ('done', 'cancel') or not move.product_uom_qty:
                 continue
-            if move.manual_consumption:
-                if move.has_tracking in ('serial', 'lot') and (not move.picked or any(not line.lot_id for line in move.move_line_ids if line.quantity and line.picked)):
-                    missing_lot_id_products += "\n  - %s" % move.product_id.display_name
+            if move.picked and move.has_tracking in ('serial', 'lot') and \
+                any(not line.lot_id for line in move.move_line_ids if line.quantity and line.picked):
+                missing_lot_id_products += "\n  - %s" % move.product_id.display_name
         if missing_lot_id_products:
             error_msg = _(
                 "You need to supply Lot/Serial Number for products and 'consume' them: %(missing_products)s",

--- a/addons/mrp/models/stock_move.py
+++ b/addons/mrp/models/stock_move.py
@@ -57,10 +57,6 @@ class StockMove(models.Model):
         help="The percentage of the final production cost for this by-product. The total of all by-products' cost share must be smaller or equal to 100.")
     product_qty_available = fields.Float('Product On Hand Quantity', related='product_id.qty_available', depends=['product_id'])
     product_virtual_available = fields.Float('Product Forecasted Quantity', related='product_id.virtual_available', depends=['product_id'])
-    manual_consumption = fields.Boolean(
-        'Manual Consumption', compute='_compute_manual_consumption', store=True, readonly=False,
-        help="When activated, then the registration of consumption for that component is recorded manually exclusively.\n"
-             "If not activated, and any of the components consumption is edited manually on the manufacturing order, Odoo assumes manual consumption also.")
 
     @api.depends('product_id.bom_ids', 'product_id.bom_ids.uom_id')
     def _compute_allowed_uom_ids(self):
@@ -75,15 +71,6 @@ class StockMove(models.Model):
             if move.production_id:
                 move.packaging_uom_id = move.production_id.uom_id
 
-    @api.depends('product_id')
-    def _compute_manual_consumption(self):
-        for move in self:
-            # when computed for new_id in onchange, use value from _origin
-            if move != move._origin:
-                move.manual_consumption = move._origin.manual_consumption
-            elif not move.manual_consumption:
-                move.manual_consumption = move._is_manual_consumption()
-
     @api.depends('raw_material_production_id.location_src_id', 'production_id.location_src_id')
     def _compute_location_id(self):
         ids_to_super = set()
@@ -215,16 +202,17 @@ class StockMove(models.Model):
     def _onchange_product_uom_qty(self):
         if self.uom_id and self.raw_material_production_id \
             and (self.has_tracking not in ['lot', 'serial']) \
-            and self.state not in ('draft', 'cancel', 'done'):
+            and self.state not in ('draft', 'cancel', 'done') and not self.picked:
             mo = self.raw_material_production_id
             new_qty = self.uom_id.round((mo.qty_producing - mo.qty_produced) * self.unit_factor)
+            self.env = self.env(context=dict(self.env.context, skip_mark_picked=True))
             self.quantity = new_qty
 
-    @api.onchange('quantity', 'uom_id', 'picked')
+    @api.onchange('quantity')
     def _onchange_quantity(self):
         if self.raw_material_production_id and self.uom_id and \
-            not self.uom_id.is_zero(self.quantity) and self.uom_id.compare(self.product_uom_qty, self.quantity) != 0:
-            self.manual_consumption = True
+            not self.env.context.get('skip_mark_picked') \
+            and self.uom_id.compare(self.product_uom_qty, self.quantity) != 0:
             self.picked = True
 
     @api.constrains('quantity', 'raw_material_production_id')
@@ -240,10 +228,8 @@ class StockMove(models.Model):
         - Moves from a copied MO
         - Backorders
         """
-        if self.env.context.get('force_manual_consumption'):
+        if self.env.context.get('force_move_picked'):
             for vals in vals_list:
-                if 'quantity' in vals:
-                    vals['manual_consumption'] = True
                 vals['picked'] = True
         mo_id_to_mo = defaultdict(lambda: self.env['mrp.production'])
         product_id_to_product = defaultdict(lambda: self.env['product.product'])
@@ -282,7 +268,7 @@ class StockMove(models.Model):
 
     def write(self, vals):
         moves_to_update = False
-        if self.env.context.get('force_manual_consumption') and 'quantity' in vals:
+        if self.env.context.get('force_move_picked') and 'quantity' in vals:
             moves_to_update = self.filtered(lambda move: move.product_uom_qty != vals['quantity'])
         if 'product_uom_qty' in vals and 'move_line_ids' in vals:
             # first update lines then product_uom_qty as the later will unreserve
@@ -292,7 +278,7 @@ class StockMove(models.Model):
         old_demand = {move.id: move.product_uom_qty for move in self}
         res = super().write(vals)
         if moves_to_update:
-            moves_to_update.write({'manual_consumption': True, 'picked': True})
+            moves_to_update.write({'picked': True})
         if 'product_uom_qty' in vals and not self.env.context.get('no_procurement', False):
             # when updating consumed qty need to update related pickings
             # context no_procurement means we don't want the qty update to modify stock i.e create new pickings
@@ -328,8 +314,34 @@ class StockMove(models.Model):
         if procurements:
             self.env['stock.rule'].run(procurements)
 
+    def _do_unreserve(self):
+        # Moves with a quantity should be protected from unreservation (similar to picked).
+        # This override is to do that for moves which got updated from updating qty_producing as
+        # they are not marked as picked any more
+        if not self.env.context.get('skip_mo_check'):
+            protected = self.filtered(
+                lambda m: m.raw_material_production_id
+                and not m.picked
+                and m.quantity
+                and m.raw_material_production_id.qty_producing > 0
+            )
+            return super(StockMove, self - protected)._do_unreserve()
+        return super()._do_unreserve()
+
     def _action_assign(self, force_qty=False):
-        res = super(StockMove, self)._action_assign(force_qty=force_qty)
+        # Moves with a quantity should be protected from re-assignation (similar to picked).
+        # This override is to do that for moves which got updated from updating qty_producing as
+        # they are not marked as picked any more
+        if not self.env.context.get('skip_mo_check'):
+            protected = self.filtered(
+                lambda m: m.raw_material_production_id
+                and not m.picked
+                and m.quantity
+                and m.raw_material_production_id.qty_producing > 0
+            )
+            res = super(StockMove, self - protected)._action_assign(force_qty=force_qty)
+        else:
+            res = super()._action_assign(force_qty=force_qty)
         for move in self.filtered(lambda x: x.production_id or x.raw_material_production_id):
             if move.move_line_ids:
                 move.move_line_ids.write({'production_id': move.raw_material_production_id.id,
@@ -394,7 +406,7 @@ class StockMove(models.Model):
             action['name'] = _("Components")
             action['views'] = [(self.env.ref('mrp.view_stock_move_operations_raw').id, 'form')]
             action['context']['show_destination_location'] = False
-            action['context']['force_manual_consumption'] = True
+            action['context']['force_move_picked'] = True
             action['context']['active_mo_id'] = self.raw_material_production_id.id
         elif self.production_id:
             action['name'] = _("Move Byproduct")
@@ -499,7 +511,6 @@ class StockMove(models.Model):
             'state': 'draft' if self.state == 'draft' else 'confirmed',
             'reservation_date': self.reservation_date,
             'date_deadline': self.date_deadline,
-            'manual_consumption': self._is_manual_consumption(),
             'move_orig_ids': [Command.link(m.id) for m in self.mapped('move_orig_ids')],
             'move_dest_ids': [Command.link(m.id) for m in self.mapped('move_dest_ids')],
             'procure_method': self.procure_method,
@@ -647,20 +658,12 @@ class StockMove(models.Model):
             }
         return res
 
-    def _is_manual_consumption(self):
-        self.ensure_one()
-        return self._determine_is_manual_consumption(self.bom_line_id)
-
-    @api.model
-    def _determine_is_manual_consumption(self, bom_line):
-        return bom_line and bom_line.operation_id
-
     def _get_relevant_state_among_moves(self):
         res = super()._get_relevant_state_among_moves()
         if res == 'partially_available'\
                 and self.raw_material_production_id\
                 and all(move.should_consume_qty and move.uom_id.compare(move.quantity, move.should_consume_qty) >= 0
-                        or (move.uom_id.compare(move.quantity, move.product_uom_qty) >= 0 or (move.manual_consumption and move.picked))
+                        or (move.uom_id.compare(move.quantity, move.product_uom_qty) >= 0 or move.picked)
                         for move in self):
             res = 'assigned'
         return res
